evo_researcher/autonolas/research.py

```diff
@@ -1144,18 +1151,29 @@ def make_prediction(prompt: str, additional_information: str, **kwargs) -> Predi
     )
     
     prediction_prompt = ChatPromptTemplate.from_template(template=PREDICTION_PROMPT)
-    prediction_chain = (
-        prediction_prompt |
-        ChatOpenAI(model="gpt-3.5-turbo-1106", openai_api_key=open_ai_key) |
-        StrOutputParser()
-    )
 
-    response = prediction_chain.invoke({
-        "user_prompt": prompt,
-        "additional_information": additional_information,
-        "timestamp": formatted_time_utc,
-    })
-    response = json.loads(clean_completion_json(response))
+    llm = ChatOpenAI(model=engine, temperature=temperature, openai_api_key=open_ai_key)
+    formatted_messages = prediction_prompt.format_messages(user_prompt=prompt, additional_information=additional_information, timestamp=formatted_time_utc)
+    generation = llm.generate([formatted_messages], logprobs=True, top_logprobs=5)
+
+    completion = generation.generations[0][0].text
+
+    # Get probability that is based on the token's top logprobs.
+    decision, probability = None, None
+    for token in generation.generations[0][0].generation_info["logprobs"]["content"]:
+        # Check if the token is a decision token, we prompt the model for it to be there, so it is in 99% of cases.
+        if token["token"] in ("y", "n"):
```

Comment:
> How do you know that the first y or n token you hit corresponds to the decision token, and not some other token?

Because of the prompt and tokenization, the first token of `y` or `n` is this decision; if it isn't, LLM provides the wrong completion, and JSON parsing will fail.

```
* The JSON must contain five fields: "decision", "p_yes", "p_no", "confidence", and "info_utility".
   - "decision": The decision you made. Either `y` (for `Yes`) or `n` (for `No`).
```

> we can have some certainty that for the tokenizer used by this model, the letters n and y map to tokens

Yeah I tested this manually, `y` and `n` there are tokens.

>  but for a different tokenizer, this might not be the case.

Yep that's true, I don't think OpenAI will change tokenization but if we plug this to another model, it can break.